AX: text-transform: full-size-kana must not affect AT/speech output

hoffmanjoshua opened the pull request:
#### 00c8f47583643ff7824ed18bd0470c1b066ee74a
<pre>
AX: text-transform: full-size-kana must not affect AT/speech output
<a href="https://bugs.webkit.org/show_bug.cgi?id=261565">https://bugs.webkit.org/show_bug.cgi?id=261565</a>
<a href="https://rdar.apple.com/115504070">rdar://115504070</a>

Reviewed by Tyler Wilcock.

`full-size-kana` is used to improve legibility issues with small Kana text. Readers are the ones
who are supposed to make the distinction, but for ATs, we need to make sure we aren&apos;t using this
visually transformed text for speech output.

This patch adds a new IteratorBehavior, IgnoresFullSizeKana, to the text iterator for use by
Accessibility. This will return the original, untransformed text, to our string-creation
methods.

* LayoutTests/accessibility/full-size-kana-untransformed-expected.txt: Added.
* LayoutTests/accessibility/full-size-kana-untransformed.html: Added.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::stringForRange const):
(WebCore::AccessibilityObject::stringForVisiblePositionRange):
(WebCore::AccessibilityObject::textIteratorBehaviorForTextRange const):
(WebCore::AccessibilityObject::textIteratorIgnoringFullSizeKana):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::AccessibilityObject::contentForRange const):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::emitText):
* Source/WebCore/editing/TextIteratorBehavior.h:

Canonical link: <a href="https://commits.webkit.org/282258@main">https://commits.webkit.org/282258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75417630cfbc494cfdccb4bfb4e29b3dae725c26

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62573 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41928 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15168 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66557 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13125 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64693 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49615 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13461 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50418 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9046 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65642 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38967 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54196 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31160 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35689 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11531 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12053 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57275 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11847 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68288 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6519 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11533 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57797 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6548 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54243 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57991 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13899 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5438 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37728 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38813 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39910 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38557 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/full-size-kana-untransformed.html:
```diff
@@ -0,0 +1,32 @@
+<script src="../resources/js-test.js"></script>
+<meta charset=utf-8>
+</head>
+<body id="body">
```
Is this `id` necessary?

hoffmanjoshua replied:
Nope (left over from a copy-and-paste). I'll remove it in the next revision.

twilco commented on LayoutTests/accessibility/full-size-kana-untransformed.html:
```diff
@@ -0,0 +1,32 @@
+</div>
+
+<script>
+var output  = "This tests that text with full-size-kana returns the untransformed text to Accessibility APIs.\n\n";
```
Extra space between `output` and `=` I think

twilco commented on Source/WebCore/editing/TextIteratorBehavior.h:
```diff
@@ -69,6 +69,9 @@ enum class TextIteratorBehavior : uint16_t {
     EmitsObjectReplacementCharactersForAttachments = 1 << 13,
 #endif
+
+    // Used by AX to expose untransformed kana text.
```
It might be good to spell out "accessibility" instead of "AX" (I've found some folks don't know that short-hand, and this change is not in the `accessibility` subfolder so might be good not to assume knowledge)

hoffmanjoshua replied:
Good idea, I can update this!

cookiecrook replied:
Yes, AX is a somewhat Apple-internal jargon, related to the AX API... Short for Accessibility in the same way that NS is short for Next Step, but how many people know that? 😅

twilco commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -1157,8 +1157,14 @@ void TextIterator::emitText(Text& textNode, RenderText& renderer, int textStartO
+    if (m_behaviors.contains(TextIteratorBehavior::IgnoresFullSizeKana)) {
+        if (renderer.style().textTransform().contains(TextTransform::FullSizeKana))
+            shouldIgnoreFullSizeKana = true;
+    }
```
Could this be written as:

```cpp
bool shouldIgnoreFullSizeKana = m_behaviors.contains(TextIteratorBehavior::IgnoresFullSizeKana) && renderer.style().textTransform().contains(TextTransform::FullSizeKana);
```

?

hoffmanjoshua replied:
Yep, I'll simplify this

cookiecrook commented on LayoutTests/accessibility/full-size-kana-untransformed-expected.txt:
```diff
@@ -0,0 +1,8 @@
+PASS successfullyParsed is true
+
+TEST COMPLETE
+じゆう must read the same as じゅう not じゆう.
```
FYI the related WPT test just landed, too.
https://github.com/web-platform-tests/wpt/pull/47535

AndresGonzalezApple commented on LayoutTests/accessibility/full-size-kana-untransformed.html:
```diff
@@ -0,0 +1,32 @@
+
+    var startMarker = textElement.startTextMarkerForTextMarkerRange(textMarkerRange);
+    var endMarker = textElement.endTextMarkerForTextMarkerRange(textMarkerRange);
+    textMarkerRange = textElement.textMarkerRangeForMarkers(startMarker, endMarker);
```
Why are we re-creating the range here?

hoffmanjoshua replied:
I was following a pattern from other tests, but I think we don't have to reconstruct. I'll remove that.

AndresGonzalezApple commented on Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
```diff
@@ -196,7 +196,7 @@ static void attributedStringAppendWrapper(NSMutableAttributedString *attrString,
+    for (TextIterator it(range, { TextIteratorBehavior::IgnoresFullSizeKana }); !it.atEnd(); it.advance()) {
         Node& node = it.range().start.container;
 
         // Non-zero length means textual node, zero length means replaced node (AKA "attachments" in AX).
```
Maybe we can add a helper function that returns a TextIterator with the behaviors we want instead of calling the constructor directly.
Also, are we covering all the TextIterators we use? e.g., the ones used to build AttributedStrings?

hoffmanjoshua replied:
I can definitely add a helper to make this cleaner.

For your second point, this covers all places where we return strings (both attributed and regular). AttributedStrings are created using `contentForRange`, which we now use this new TextIteratorBehavior in. We don't need to use this behavior when computing text markers with iterators, however, since this only affects character appearance.

hoffmanjoshua committed:
AX: text-transform: full-size-kana must not affect AT/speech output
https://bugs.webkit.org/show_bug.cgi?id=261565
rdar://115504070

Reviewed by Tyler Wilcock.

`full-size-kana` is used to improve legibility issues with small Kana text. Readers are the ones
who are supposed to make the distinction, but for ATs, we need to make sure we aren't using this
visually transformed text for speech output.

This patch adds a new IteratorBehavior, IgnoresFullSizeKana, to the text iterator for use by
Accessibility. This will return the original, untransformed text, to our string-creation
methods.

* LayoutTests/accessibility/full-size-kana-untransformed-expected.txt: Added.
* LayoutTests/accessibility/full-size-kana-untransformed.html: Added.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::stringForRange const):
(WebCore::AccessibilityObject::stringForVisiblePositionRange):
(WebCore::AccessibilityObject::textIteratorBehaviorForTextRange const):
(WebCore::AccessibilityObject::textIteratorIgnoringFullSizeKana):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::AccessibilityObject::contentForRange const):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::emitText):
* Source/WebCore/editing/TextIteratorBehavior.h:

Canonical link: https://commits.webkit.org/282258@main
